### PR TITLE
020 Script for requirement [Policies] [External UCS] SDL informs HMI about <externalConsentStatus> via GetListOfPermissions response

### PR DESCRIPTION
**Pull request by vvvakulenko**
020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua
test is a part of TS #1165 
@istoimenova Please review

### CODE REVIEW

File: test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+--[[ Test ]]
+commonFunctions:newTestCasesGroup("Test")
+
+function Test:TestStep_GetListofPermissions_entityID_invalid()
+  local RequestIdListOfPermissions = self.hmiConnection:SendRequest("SDL.GetListOfPermissions", {appID = self.applications[config.application1.registerAppInterfaceParams.appName]})
+
+  EXPECT_HMIRESPONSE(RequestIdListOfPermissions, {
+    code = "0",
+    allowedFunctions = {
+    { name = "Location-1", id = 156072572},
```
Comment (istoimenova) [inline test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua]: @vvvakulenko, name should be "Location"

File: test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            :Do(function(_,_)
+                self.hmiConnection:SendNotification("SDL.OnAppPermissionConsent",
+                  {
+                    appID = self.applications[config.application1.registerAppInterfaceParams.appName],
+                    consentedFunctions = {
+                      { allowed = true, id = 156072572, name = "Location-1"},
+                      { allowed = true, id = 1809526495, name = "Notifications"}
+                    },
+                    externalConsentStatus = {
+                      {entityType = 13, entityID = "invalidID", status = "ON"}
```
Comment (istoimenova) [inline test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua]: @vvvakulenko, entityID may have different invalid verification.
You may use loop with defining them, as example:
local params_invalid_data =
{
{param_value = "invalidValue", comment = "String"}
{param_value = 1.32 comment = "Float"},
{param_value = {}, comment = "Empty table" },
{param_value = { entityType = 1, entityID = 1 }, comment = "Non-empty table"},
{param_value = xxxx, comment = "OutOfLowerBound"},
{param_value = xxxx, comment = "OutOfUpperBound" },
{param_value = "", comment = "Empty" },
{param_value = nil, desc = "Null" }
}

.......................
for i = 1, #params_invalid_data do
............
externalConsentStatus = { {entityType = 13, entityID = params_invalid_data[i], status = "ON"} }
.............
end

File: test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                    },
+                    source = "GUI"
+                  })
+                EXPECT_NOTIFICATION("OnPermissionsChange"):Times(0)
+                commonTestCases:DelayedExp(10000)
```
Comment (istoimenova) [inline test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua]: @vvvakulenko, please declare commonTestCases

File: test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua (hunk 2 above)
Comment (vvvakulenko) [inline test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua]: @istoimenova Please see updates in 17c7bfeb7147aeb4f4c46b8867936cdf4695876a

File: test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua (hunk 3 above)
Comment (vvvakulenko) [inline test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua]: @istoimenova Please see updates in 17c7bfeb7147aeb4f4c46b8867936cdf4695876a

File: test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua (hunk 1 above)
Comment (vvvakulenko) [inline test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua]: @istoimenova Please see updates in 17c7bfeb7147aeb4f4c46b8867936cdf4695876a

File: test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+--[[ Preconditions ]]
+commonFunctions:newTestCasesGroup("Preconditions")
+
+function Test:Precondition_trigger_getting_device_consent()
+  testCasesForPolicyTable:trigger_getting_device_consent(self, config.application1.registerAppInterfaceParams.appName, config.deviceMAC)
+end
+
+for i = 1, #params_invalid_data do
+  Test["Precondition_PTU_and_OnAppPermissionConsent_Invalid_"..params_invalid_data[i]] = function(self)
+    local ptu_file_path = "files/jsons/Policies/Related_HMI_API/"
```
Comment (istoimenova) [inline test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua]: @vvvakulenko, To have correct PTU and OnAppPermissionChange notification sent after each PTU, you should change permission in ptu file. So you need as example:

local ptu_file
local consentedgroups
local allowed_func

if ( (i % 2) == 0 ) then
ptu_file = "OnAppPermissionConsent_ptu1.json"
consentedgroups = {{ allowed = true, id = 4734356, name = "DrivingCharacteristics-3"}}
allowed_func = { { name = "DrivingCharacteristics", id = 4734356}}
else
ptu_file = "OnAppPermissionConsent_ptu.json"
consentedgroups = {
{ allowed = true, id = 156072572, name = "Location-1"},
{ allowed = true, id = 1809526495, name = "Notifications"}
}

  allowed_func = {
    { name = "Location", id = 156072572},
    { name = "Notifications", id = 1809526495}
  }

end

File: test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    
+    testCasesForPolicyTable:flow_SUCCEESS_EXTERNAL_PROPRIETARY(self, nil, nil, nil, ptu_file_path, nil, ptu_file)
+    
+    EXPECT_HMINOTIFICATION("SDL.OnAppPermissionChanged",{ appID = self.applications[config.application1.registerAppInterfaceParams.appName]})
+    :Do(function(_,data)
+      if (data.params.appPermissionsConsentNeeded== true) then
+        local RequestIdListOfPermissions = self.hmiConnection:SendRequest("SDL.GetListOfPermissions", {appID = self.applications[config.application1.registerAppInterfaceParams.appName]})
+          EXPECT_HMIRESPONSE(RequestIdListOfPermissions,{result = {code = 0, method = "SDL.GetListOfPermissions",
+            -- allowed: If ommited - no information about User Consent is yet found for app.
+            allowedFunctions = {
```
Comment (istoimenova) [inline test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua]: @vvvakulenko, based on comment on line 80, you should update

allowedFunctions = = allowed_func,

File: test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+          local ReqIDGetUserFriendlyMessage = self.hmiConnection:SendRequest("SDL.GetUserFriendlyMessage",
+          {language = "EN-US", messageCodes = {"AppPermissions"}})
+          
+          EXPECT_HMIRESPONSE(ReqIDGetUserFriendlyMessage,
+          {result = {code = 0, messages = {{messageCode = "AppPermissions"}}, method = "SDL.GetUserFriendlyMessage"}})
+          :Do(function(_,_)
+            self.hmiConnection:SendNotification("SDL.OnAppPermissionConsent",
+            {
+              appID = self.applications[config.application1.registerAppInterfaceParams.appName],
+              consentedFunctions = {
```
Comment (istoimenova) [inline test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua]: @vvvakulenko, based on comment on line 80, you should update:
consentedFunctions = consentedgroups

File: test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    end)
+  end
+end
+
+--[[ Test ]]
+commonFunctions:newTestCasesGroup("Test")
+
+function Test:TestStep_GetListofPermissions_entityID_invalid()
+  local RequestIdListOfPermissions = self.hmiConnection:SendRequest("SDL.GetListOfPermissions", {appID = self.applications[config.application1.registerAppInterfaceParams.appName]})
+
```
Comment (istoimenova) [inline test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua]: @vvvakulenko, please update based on updates in previous step as:

local allowed_func

if ( (i % 2) == 0 ) then
allowed_func = { { name = "DrivingCharacteristics", id = 4734356}}
else
allowed_func = {
{ name = "Location", id = 156072572},
{ name = "Notifications", id = 1809526495}
}
end

File: test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+--[[ Test ]]
+commonFunctions:newTestCasesGroup("Test")
+
+function Test:TestStep_GetListofPermissions_entityID_invalid()
+  local RequestIdListOfPermissions = self.hmiConnection:SendRequest("SDL.GetListOfPermissions", {appID = self.applications[config.application1.registerAppInterfaceParams.appName]})
+  
+  EXPECT_HMIRESPONSE(RequestIdListOfPermissions, {
+    code = "0",
+    allowedFunctions = {
```
Comment (istoimenova) [inline test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua]: @vvvakulenko, based on comment on line 124, you should use:

allowedFunctions = allowed_func

**Comment by vvvakulenko** [comment]
@istoimenova Please see updated in 403486902c7c515cec0b1135cfc9637a4aa99d47

File: test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+for i = 1, #params_invalid_data do
+  if(i > 1) then
+    commonFunctions:newTestCasesGroup("Preconditions")
+
+    Test["Precondition_trigger_user_request_update_from_HMI_"..params_invalid_data[i].comment] = function(self)
+      testCasesForPolicyTable:trigger_user_request_update_from_HMI(self)
+    end
+  end
+
+  Test["Precondition_PTU_and_OnAppPermissionConsent_Invalid_"..params_invalid_data[i]] = function(self)
```
Comment (istoimenova) [inline test_scripts/Polices/External_UCS_Informing_HMI/020_ATF_N_GetListofPermissions_empty_due_to_invalid_param_entityID_OnAppPermissionChange.lua]: @vvvakulenko, you should concatenate params_invalid_data[i].comment

**Pull request merged**